src/marqo/tensor_search/telemetry.py

```diff
@@ -0,0 +1,214 @@
+from starlette.middleware.base import BaseHTTPMiddleware
+from starlette.requests import Request
+from starlette.responses import Response
+from typing import Any, Callable, Dict, List, Optional, Union
+import json
+from collections import defaultdict
+from contextlib import contextmanager
+import time
+from contextvars import ContextVar
+from marqo.tensor_search.models.add_docs_objects import AddDocsParams
+from marqo.tensor_search.tensor_search_logging import get_logger
+
+logger = get_logger(__name__)
+
+
+class TimerError(Exception):
+    """An error occured whn operating on a metric timer (e.g. stopping a stopped timer)"""
+    pass
+
+
+class Timer:
+    def __init__(self):
+        self.start_time = None
+
+    def start(self) -> None:
+        """Start a new timer"""
+        if self.start_time is not None:
+            logger.warn(f"'.start()' called on already running timer.")
+        else:
+            self.start_time = time.perf_counter()
```

Comment:
Because we're wanting to return it as milliseconds, using time.perf_counter_ns would force us to multiply the time by 1e-6, leading to a much higher precision loss ~= 6 less digits (e.g. 5597.397250006907 vs 4816.107667).